wandb/sdk/internal/tb_watcher.py

```diff
@@ -390,23 +400,53 @@ class TBHistory(object):
 
     def __init__(self) -> None:
         self._step = 0
+        self._step_size = 0
         self._data = dict()
         self._added = []
 
     def _flush(self) -> None:
         if not self._data:
             return
+        # A single tensorboard step may have too much data
+        # we just drop the largest keys in the step if it does.
+        # TODO: we could flush the data across multiple steps
+        if self._step_size > file_stream.MAX_LINE_SIZE:
+            metrics = [(k, sys.getsizeof(v)) for k, v in self._data.items()]
+            metrics.sort(key=lambda t: t[1], reverse=True)
+            bad = 0
+            dropped_keys = []
+            for k, v in metrics:
+                if self._step_size - bad < file_stream.MAX_LINE_SIZE:
+                    break
+                else:
+                    bad += v
+                    dropped_keys.append(k)
+                    del self._data[k]
+            wandb.termwarn(
+                "Step {} exceeds max data limit, dropping {} of the largest keys:".format(
+                    self._step, len(dropped_keys)
+                )
+            )
+            print("\t" + ("\n\t".join(dropped_keys)))
```

Comment:
This was the error that Graphcore was seeing, atleast now we tell the user exactly what keys we're dropping if a single "row" is more than 4 megabytes.